### DISCUSSION TITLE
[Contract Verification] Factory automatic verification

### DISCUSSION THREAD

**Discussion by nfts2me** [zkSync-Community-Hub/zksync-developers#72] [Contract Verification] [answered]
### Environment

Mainnet

### Issue Type

- [ ] Plugin
- [X] Block Explorer

### Contract Address

Factory

### Additional Details

I have a contract factory which creates new contracts using their bytecode hash to create them. 

How to get those contracts created with the factory automatically verified?
All of them share the same source code, it would make sense to verify all of them with the same bytecode automatically despite having different addresses and constructor arguments.

**Comment by idea404** [comment]
A suggestion would be to have a machine monitor whenever the factory deploys a new contract and then programatically perform the verification. Not the neatest setup but I do not think you can upload a script to the explorer that would perform this step for you whenever the factory deploys a contract.

As for standard functionality in the explorer, the team is exploring ways to support this in future once the verification functionality has been implemented to the app.

**Reply by nfts2me** [reply]
On other explorers from other EVMs, proxies are automatically verified with the source of the destination contract.
Here, it would be even more straightforward, as the code is the same for all the contracts.
All the contracts with the same sourcecode should be automatically verified from the zkSync explorer.

**Comment by lysumeduso** [comment]
To automatically verify contracts created by the factory, you can use a contract verification service or tool that supports automatic verification based on bytecode hash. This way, you can submit the bytecode hash of the source code used by the factory to the verification service, and it will automatically verify any contracts created with that bytecode hash. This approach ensures that all contracts created by the factory are verified without the need for manual intervention for each contract address or constructor argument combination.

**Reply by nfts2me** [reply]
What's that?
I've never heard of a "contract verification service" on zkSync, just this: https://docs.zksync.io/build/tooling/block-explorer/contract-verification.html

**Reply by nfts2me** [reply]
OK, I see. Just a IA bot response farming. Please, any @admin remove this.

**Comment by fitzgibbonsdustin** [comment]
To automatically verify contracts created by a factory contract with the same source code, you can follow these steps:

Use a deployment script or a deployment function in your factory contract to deploy new contracts.
After deploying each new contract, retrieve its bytecode hash.
Use a plugin or external tool (if available) to automatically submit verification requests for each contract using its bytecode hash.
Ensure that the verification requests include the source code of the contract, constructor arguments (if any), and other necessary information.
Monitor the verification status to ensure that all contracts are successfully verified.
By automating the verification process in this way, you can ensure that all contracts created by the factory contract are automatically verified with the same bytecode, regardless of their addresses or constructor arguments.

**Reply by nfts2me** [reply]
Another IA bot response farming. Please, any @admin remove this.

**Comment by phicurby** [comment]
To automatically verify contracts created by a factory, you can use a contract verification service or tool that supports automatic verification based on bytecode hash. By providing the bytecode hash of the source contract, the verification service can verify all contracts created by the factory that share the same source code. Ensure that the verification service you choose supports this feature and follow their documentation or guidelines to automate the verification process.

**Reply by nfts2me** [reply]
Another IA bot response farming. Please, any @admin remove this.

**Comment by camicagochristie** [comment]
To automatically verify contracts created by a factory, you can use a verification plugin or script that supports automatic verification based on bytecode. Ensure that the plugin or script is configured to verify contracts with the same source code but different addresses and constructor arguments.

**Comment by AlbionaHoti** [comment]
Hi @nfts2me,

It's been a long time, but I just wanted to share that now we have Etherscan support for the zkSync Era network; please find the link [here](https://era.zksync.network/), and you will be able to verify factory SC automatically there.

**Reply by nfts2me** [reply]
Thank you @AlbionaHoti , but unfortunately it is not working.

Last time we checked (a few days ago) latest zksolc was not supported, and hardhat-verify method was not working either.
But not only that, trying on an older version, there were only error, it didn't compile using the UI.

Do you know anyone from the team to talk via Telegram or other way to try to solve it?

**Reply by bxpana** [reply]
What version of Solidity is your factory in? When you say they aren't working what errors are you seeing?

**Reply by nfts2me** [reply]
solidity 0.8.24
zksolc 1.4 (which is over a month old)

Right now, for instance, the latest 0.8.5 solidity compiler is shown, but not the latest zksold version.
<img width="347" alt="image" src="https://github.com/zkSync-Community-Hub/zksync-developers/assets/109629895/3fb306c4-8e23-4204-b7c7-d6f9c4a879f3">

Error when trying using the latest version through the UI (tried with 1.3.22 compiled contract).
```
Compiler debug log:
 Error! Unable to generate Contract Bytecode and ABI (General Exception, unable to get compiled [bytecode])
 For troubleshooting, you can try compiling your source code with the Remix - Solidity IDE and check for exceptions

```  
![telegram-cloud-photo-size-4-6019579893400059946-y](https://github.com/zkSync-Community-Hub/zksync-developers/assets/109629895/0623ce63-bcb0-426c-95dc-f07c57576a21)

I also haven't been able to verify using hardhat-verify plugin for the era.zksync.network explorer.

**Reply by bxpana** [reply]
Apologies for the delay here, catching up after being out sick. I'll report back to the team and see what's up. 

You're not able to verify using our explorer, https://explorer.zksync.io/ either correct?

**Comment by glasrudmary** [comment]
To automatically verify all contracts created by a factory with the same source code and different addresses and constructor arguments, you can use a contract verification service that supports automatic verification. These services typically use the bytecode hash to identify the contract and verify it against the corresponding source code. Some popular contract verification services include Etherscan, Sourcify, and Truffle Verify. By using one of these services, you can ensure that all contracts created by the factory are automatically verified without needing to manually verify each one.

**Reply by nfts2me** [reply]
Another farmer bot...

**Comment by nfts2me** [comment]
The problem with that explorer is that we use a contract factory.
And contracts are created using create2 from the bytecodehash.
So we cannot verify all of them individually as we are a permissionless
protocol.

**Reply by bxpana** [reply]
Ok let me look into this

**Reply by bxpana** [reply, marked as answer]
So it looks like this is currently a known issue of not being able to verify contracts that have been deployed from another contract. It's something we have planned to implement in the future though. Is that a blocker for you guys right now?

**Reply by nfts2me** [reply]
Rather than a blocker, it is a complain from the users that are unable to see their contracts verified.

**Reply by bxpana** [reply]
Understood. We apologize for the inconvenience, but will keep the community updated on when we have that feature available.

**Reply by gepng** [reply]
@bxpana any update on this?

**Reply by bxpana** [reply]
@gedithejedi no update on this and will probably be a bit before we have this as an option.
